pyfstat/snr.py

```diff
@@ -0,0 +1,230 @@
+        return expected_2F, stdev_2F
+
+    def compute_Mmunu(self, Alpha, Delta):
+
+        sky = lal.SkyPosition()
+        sky.longitude = Alpha
+        sky.latitude = Delta
+        sky.system = lal.COORDINATESYSTEM_EQUATORIAL
+        lal.NormalizeSkyPosition(sky.longitude, sky.latitude)
+
+        # FIXME Missing Mmunu.Sinv_Tsft = t_sft / psd !!!!!!
+        return lalpulsar.ComputeMultiAMCoeffs(
+            multiDetStates=self.detector_states,
+            multiWeights=self.noise_weights,
+            skypos=sky,
+        ).Mmunu
+
+    def _convert_amplitude_parameters(self, h0, cosi, aPlus, aCross):
+        h0_cosi = h0 is not None and cosi is not None
+        aPlusCross = aPlus is not None and aCross is not None
+
+        if h0_cosi == aPlusCross:
+            raise ValueError("Need either (h0, cosi) or (aPlus, aCross), but not both")
+
+        if h0_cosi:
+            aPlus = 0.5 * h0 * (1 + cosi**2)
+            aCross = h0 * cosi
+            return aPlus, aCross
+
+        return aPlus, aCross
+
+
+class DetectorStates:
+    def __init__(self):
+        self.ephems = lalpulsar.InitBarycenter(*get_ephemeris_files())
+
+    def multi_detector_states(self, timestamps, detectors=None, time_offset=0):
+        """
+        Parameters
+        ----------
+        time_stamps: array-like or dict
+            GPS timestamps at which detector states will be retrieved.
+            If array, use the same set of timestamps for all detectors,
+            which must be explicitly given by the user via`detectors`.
+            If dictionary, each key should correspond to a valid detector name
+            to be parsed by XLALParseMultiLALDetector and the associated value
+            should be an array-like set of GPS timestamps for each individual detector.
+
+        detectors: list[str] or comma-separated string
+            List of detectors to be parsed using XLALParseMultiLALDetector.
+
+        time_offset: float
+            Timestamp offset to retrieve detector states.
+
+
+        Returns
+        -------
+        multi_detector_states: lalpulsar.MultiDetectorStateSeries
+            Resulting multi-detector states produced by XLALGetMultiDetectorStates
+        """
+
+        self._parse_timestamps_and_detectors(timestamps, detectors)
+        return lalpulsar.GetMultiDetectorStates(
+            self.multi_timestamps,
+            self.multi_detector,
+            self.ephems,
+            time_offset,
+        )
+
+    def multi_detector_states_from_sfts(
+        self,
+        sftfilepath,
+        central_frequency,
+        frequency_wing_bins=1,
+        time_offset=0,
+        sft_constraint=None,
+        return_sfts=False,
+    ):
+        multi_sft_catalog = lalpulsar.GetMultiSFTCatalogView(
+            lalpulsar.SFTdataFind(sftfilepath, sft_constraint)
+        )
+        df = multi_sft_catalog.data[0].data[0].header.deltaF
+        multi_sfts = lalpulsar.LoadMultiSFTsFromView(
```

Comment:
SegFault is in this line

Reply:
I already encountered this somewhere: Catalog view is a thin wrapper around SFTCatalog, meaning it does *not* own any memory. SWIG, however, seems not to know that (yet), hence it tries to free the same memory twice.

Probably the following SWIG directive should be added to `LoedMultiSFTsFromView`:
```
#ifdef SWIG // SWIG interface directives
SWIGLAL( RETURN_OWNED_BY_1ST_ARG( <first argument>*, XLALLoadMultiSFTsFromView) );
#endif
```
Not sure rn since it's been a while, but I'll leave it here for the record.

@dbkeitel